admin/aws.py

```diff
@@ -0,0 +1,203 @@
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+"""
+Effectful interface to boto.
+"""
+
+from characteristic import attributes, Attribute
+from effect import sync_performer, TypeDispatcher
+
+import boto
+
+
+@attributes([
+    "bucket",
+    "prefix",
+    "target_prefix",
+])
+class UpdateS3RoutingRule(object):
+    """
+    Update a routing rule for an S3 bucket website endpoint to point to a new
+    path.
+
+    If the path is changed, return the old path.
+
+    :ivar bytes bucket: Name of bucket to change routing rule for.
+    :ivar bytes prefix: Prefix to change routing rule for.
+    :ivar bytes target_prefix: Target prefix to redirect to.
+    """
+
+
+@sync_performer
```

Comment:
The weakness of this testing strategy is that the real side-effects aren't tested. The cost of automating testing this code may be prohibitive, of course, so not necessarily worth doing.

Would it be possible to create a performer API layer that is lower-level, basically boto-level API? That way your in-memory tests would be testing much more of the logic. The boto API may be complex enough that again it's not worth the effort given this isn't core feature in the product.